lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -451,8 +451,20 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
 
     performTypoCorrection(DC, UDRE->getRefKind(), Type(), Name, Loc,
                           lookupOptions, Lookup);
+    
+    // Check if any lookups share a basename. If that's the case, we need
+    // to disambiguate the diagnostic by printing the UDRE's full name.
```

Comment:
Since a DeclName can represent either a simple or compound name, I think it'd be better to keep using DeclName as the diagnostic argument rather than printing to strings.